### ISSUE TITLE
treat identifiability by requiring centered feature functions

### ISSUE THREAD

**Issue by dswah** [dswah/pyGAM#24]
pg 164 in Wood introduces this concept. elaborated in appendix A.5

**Comment by dswah** [comment]
some more posts
https://stats.stackexchange.com/questions/110472/how-exactly-is-the-sum-or-mean-centering-constraint-for-splines-also-w-r-t-g?rq=1
https://stats.stackexchange.com/questions/33366/centering-constraints-in-mgcv-gam?rq=1

**Comment by ashrafabdul** [comment]
@dswah From what I understand, when using regression splines this can be enforced by setting the intercept parameter of the shape function to zero. However, I am not sure how this can be achieved with bsplines.
I wonder if this can this be done post hoc. Once we have the fitted shape function, can we shift each function by some constant value such that expectation of the shape function over the covariate is close to zero. We could then add all the individual shifts to the global intercept term and finally reevaluate the model performance. What do you think about this?

**Comment by Pratham-ja** [comment]
Hi @dswah,

I’ve been going through the mathematical details and the relevant literature (including the references you mentioned) to better understand the identifiability and centering constraints in GAMs.

Since this issue was last actively discussed several years ago, I wanted to check whether addressing it is still of interest for pyGAM. If so, I’d be happy to explore possible implementation approaches and share a proposal for feedback.

**Comment by DEEP-600** [comment]
Hey @dswah, been reading through this issue and the references. Here's where I've landed.
Without centering, spline basis columns have non-zero means making them linearly dependent with the intercept. PIRLS ends up with a nearly singular system, coefficients become unstable, and p-values are directly affected, I traced it through _compute_p_value at L1241, the score is a quadratic form in coef_ and its covariance, so any instability there flows straight into the p-value. Interestingly I noticed there's already a coef -= coef.mean() at L1282 inside the p-value calculation itself looks like a temporary patch for exactly this problem.

For implementation I'm thinking a center() method in the base Term class in terms.py, applied after build_columns() subtract column means from the basis. The critical design piece is predict(), means must be computed on training data during fit(), stored as fitted state, and reapplied consistently on new data.

Before touching the internals I'm planning to either put together a standalone prototype showing the effect on coefficient stability, or write up a design document covering implementation approach, predict() handling, and test strategy. Which would be more useful to you, or do you have a different direction in mind?

**Comment by sun-9545sunoj** [comment]
Hi @dswah (Daniel),

My name is **Sunoj,** and I’m a Computer Science student currently exploring projects for GSoC 2026. I’ve been following the development of pyGAM and am very interested in tackling Issue #24 regarding centered feature functions.

I’ve been reviewing the mathematical foundations in Simon Wood’s text (specifically pg. 164 and Appendix A.5). I understand that enforcing a sum-to-zero constraint on the spline basis functions is critical for model identifiability, especially to prevent collinearity between the smooth terms and the intercept.

Why I’m interested:
I have a strong background in scientific computing and open-source contributions. I’ve previously worked on:

TensorFlow/PyTorch: Resolving bugs in linear algebra functions (like tf.linalg.logdet) and handling PR workflows.

High-Performance Computing: Working with the ROCm/HIP stack on AMD GPUs.

Quantum Simulation: Implementing grid creation fixes for large-scale simulations in the qsim framework.

My Initial Thoughts on Implementation:
I plan to implement the centering logic within the base Term class in pygam/terms.py. My focus will be on using a projection matrix (likely via QR decomposition) to transform the basis functions while ensuring numerical stability during the IRLS estimation process. I also aim to include a comprehensive suite of pytest cases to verify that the condition number of the model matrix improves with these constraints.

I would love to discuss any specific preferences you have for the API structure or any edge cases (like tensor products) I should prioritize.

Looking forward to hearing your thoughts!

Best regards,
Sunoj

**Comment by roushnisareen** [comment]
Hi @dswah 
I explored the identifiability issue in spline-based terms.
My understanding is that lack of centering causes collinearity with the intercept.
I propose adding a center() method in the base Term class that ensures basis functions have zero mean over the dataset.
This can be applied during build_columns() or design matrix construction.
I’d love feedback before implementing.

**Comment by dswah** [comment]
@DEEP-600 yes precisely my thinking. i have a feature branch that attempts to perform this centering like you said. let me push it to the repo and share it with you. i think this is a fine idea
